File: objectivec/day10_part2_2016.m
#import <Foundation/Foundation.h>

@interface Bot : NSObject
@property (nonatomic, strong) NSString *lowTo, *highTo;
@property (nonatomic, strong) NSMutableArray<NSNumber *> *chips;
@end

@implementation Bot
- (instancetype)init {
    self = [super init];
    if (self) {
        _chips = [NSMutableArray array];
    }
    return self;
}
@end

int main(int argc, const char * argv[]) {
    @autoreleasepool {
        NSMutableDictionary<NSString *, Bot *> *bots = [NSMutableDictionary dictionary];
        NSMutableDictionary<NSString *, NSNumber *> *outputs = [NSMutableDictionary dictionary];
        
        NSString *filePath = @"input.txt";
        NSString *input = [NSString stringWithContentsOfFile:filePath encoding:NSUTF8StringEncoding error:nil];
        NSArray<NSString *> *lines = [input componentsSeparatedByString:@"\n"];
        
        NSRegularExpression *valueRegex = [NSRegularExpression regularExpressionWithPattern:@"value (\\d+) goes to (bot \\d+)" options:0 error:nil];
        NSRegularExpression *givesRegex = [NSRegularExpression regularExpressionWithPattern:@"(bot \\d+) gives low to (bot \\d+|output \\d+) and high to (bot \\d+|output \\d+)" options:0 error:nil];
        
        for (NSString *line in lines) {
            if ([line length] == 0) continue;
            NSTextCheckingResult *valueMatch = [valueRegex firstMatchInString:line options:0 range:NSMakeRange(0, [line length])];
            if (valueMatch) {
                NSInteger value = [[line substringWithRange:[valueMatch rangeAtIndex:1]] integerValue];
                NSString *botID = [line substringWithRange:[valueMatch rangeAtIndex:2]];
                Bot *bot = bots[botID] ?: [[Bot alloc] init];
                [bot.chips addObject:@(value)];
                bots[botID] = bot;
            } else {
                NSTextCheckingResult *givesMatch = [givesRegex firstMatchInString:line options:0 range:NSMakeRange(0, [line length])];
                if (givesMatch) {
                    NSString *botID = [line substringWithRange:[givesMatch rangeAtIndex:1]];
                    NSString *lowTo = [line substringWithRange:[givesMatch rangeAtIndex:2]];
                    NSString *highTo = [line substringWithRange:[givesMatch rangeAtIndex:3]];
                    Bot *bot = bots[botID] ?: [[Bot alloc] init];
                    bot.lowTo = lowTo;
                    bot.highTo = highTo;
                    bots[botID] = bot;
                }
            }
        }

        while (YES) {
            BOOL action = NO;
            for (Bot *b in bots.allValues) {
                if (b.chips.count == 2) {
                    action = YES;
                    NSInteger low = MIN([b.chips[0] integerValue], [b.chips[1] integerValue]);
                    NSInteger high = MAX([b.chips[0] integerValue], [b.chips[1] integerValue]);
                    [b.chips removeAllObjects];
                    if ([b.lowTo hasPrefix:@"bot"]) {
                        Bot *lowBot = bots[b.lowTo] ?: [[Bot alloc] init];
                        [lowBot.chips addObject:@(low)];
                        bots[b.lowTo] = lowBot;
                    } else {
                        outputs[b.lowTo] = @(low);
                    }
                    if ([b.highTo hasPrefix:@"bot"]) {
                        Bot *highBot = bots[b.highTo] ?: [[Bot alloc] init];
                        [highBot.chips addObject:@(high)];
                        bots[b.highTo] = highBot;
                    } else {
                        outputs[b.highTo] = @(high);
                    }
                }
            }
            if (!action) break;
        }

        NSInteger result = [outputs[@"output 0"] integerValue] * [outputs[@"output 1"] integerValue] * [outputs[@"output 2"] integerValue];
        NSLog(@"%ld", (long)result);
    }
    return 0;
}

File: fsharp/day10_part2_2016.fs

open System
open System.IO
open System.Text.RegularExpressions

[<EntryPoint>]
let main _ =
    let maxBots = 256
    let maxOutputs = 256
    let lowTo = Array.zeroCreate<int> maxBots
    let highTo = Array.zeroCreate<int> maxBots
    let chips = Array.init maxBots (fun _ -> Array.zeroCreate<int> 2)
    let chipCnt = Array.zeroCreate<int> maxBots
    let outputs = Array.zeroCreate<int> maxOutputs

    let give (target:int) (isBot:bool) (value:int) =
        if isBot then
            let c = chipCnt.[target]
            if c < 2 then
                chips.[target].[c] <- value
                chipCnt.[target] <- c + 1
        else
            outputs.[target] <- value

    let lines = File.ReadAllLines "input.txt"
    let reVal = Regex(@"value (\d+) goes to bot (\d+)")
    let reBot = Regex(@"bot (\d+) gives low to (bot|output) (\d+) and high to (bot|output) (\d+)")

    for line in lines do
        if reVal.IsMatch line then
            let m = reVal.Match line
            let v = int m.Groups.[1].Value
            let b = int m.Groups.[2].Value
            give b true v
        elif reBot.IsMatch line then
            let m = reBot.Match line
            let b = int m.Groups.[1].Value
            let lt = m.Groups.[2].Value
            let lId = int m.Groups.[3].Value
            let ht = m.Groups.[4].Value
            let hId = int m.Groups.[5].Value
            lowTo.[b]  <- if lt = "bot" then lId else -1 - lId
            highTo.[b] <- if ht = "bot" then hId else -1 - hId

    let mutable changed = true
    while changed do
        changed <- false
        for i = 0 to maxBots-1 do
            if chipCnt.[i] = 2 then
                changed <- true
                let a = chips.[i].[0]
                let b = chips.[i].[1]
                let low = if a < b then a else b
                let high = if a > b then a else b
                let lt = lowTo.[i]
                let ht = highTo.[i]
                give (if lt < 0 then -1 - lt else lt) (lt >= 0) low
                give (if ht < 0 then -1 - ht else ht) (ht >= 0) high
                chipCnt.[i] <- 0

    let result = outputs.[0] * outputs.[1] * outputs.[2]
    printfn "%d" result
    0
